dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1052 @@
+
+* Create a new https://github.com/temporalio/sdk-dotnet-incubating repo for the very first commits
+* After that repo has some code and API doc gen capability:
+  * Rename https://github.com/temporalio/sdk-dotnet to https://github.com/temporalio/legacy-dotnet (note the
+    intentionally absent "sdk" term for SEO reasons)
+    * Add README note on https://github.com/temporalio/legacy-dotnet pointing to
+      https://github.com/temporalio/sdk-dotnet
+    * Disable GH pages
+    * Archive it
+    * Consider deletion at some point in the future
+  * Rename https://github.com/temporalio/sdk-dotnet-incubating to https://github.com/temporalio/sdk-dotnet
+  * Update DNS for https://dotnet.temporal.io/ to point to vercel-based API docs page
+
+Doing the repo rename was deemed easier than going through manually closing issues, closing PRs, creating empty `main`
+branch, and keeping `master` branch around for those reading older code.
+
+## Namespace/Package
+
+* Will publish as `Temporalio`
+  * Need to reserve prefix `Temporalio` on NuGet, see
+    https://learn.microsoft.com/en-us/nuget/nuget-org/id-prefix-reservation
+* Namespace will be `Temporal`
+  * See discussion topics later about this being `Temporal.Sdk` instead
+
+## Workflow Definition
+
+Here's a definition for an other-language-defined workflow (from Python README):
+
+```csharp
+namespace MyNamespace;
+
+using System.Text.Json.Serialization;
+using Temporal.Workflow;
+
+public record GreetingInfo
+{
+    [JsonPropertyName("salutation")]
+    public string Salutation { get; init; } = "Hello";
+
+    [JsonPropertyName("name")]
+    public string Name { get; init; } = "<unknown>";
+}
+
+[Workflow]
+public interface IGreetingWorkflow
+{
+    static readonly IGreetingWorkflow Ref = Refs<IGreetingWorkflow>.Instance;
+
+    [WorkflowRun]
+    Task<string> RunAsync(GreetingInfo info);
+
+    [WorkflowSignal]
+    Task UpdateInfoAsync(GreetingInfo info);
+
+    [WorkflowSignal]
+    Task CompleteWithGreetingAsync();
+
+    [WorkflowQuery]
+    string CurrentGreeting();
+}
+```
+
+Notes:
+
+* We use regular .NET JSON serialization (see data conversion later)
+* Workflow attributes, and in the future workflow runtime features, are in the `Temporal.Workflow` namespace
+* We only support workflow interfaces right now because we don't have a workflow implementation yet
+  * For now, everything non-static must have some form of `[Workflow` attribute, and not have a default implementation
+* Every workflow interface must have the non-inheritable `Workflow` attribute
+  * Optional `Name` positional property can be given. Default is the unqualified type name with the "I" prefix removed
+    if it's an interface (which it has to be currently) and the next letter is also capital
+    * So what does `IDWorkflow` become? Yes, `DWorkflow` :-( Weighed as an acceptably rare cost for obvious benefit
+* Every workflow must explicitly have one and only one method with a non-inheritable `WorkflowRun` attribute
```

Comment:
In these languages requiring a certain method name w/ no compile-time limitation is not usually how you see these implemented. If you want to mark a method for a purpose, you use the language's preferred way of marking it. Otherwise, you use inheritance/interfaces to define method name requirements.

The benefits for inheritance are not necessarily more than you could get requiring people use a certain method name. It's more about code legibility for inheritance at that point. But yes, you can say "every workflow must have a `RunAsync` method on that class" just as easily as you could say "every workflow must have a method marked `WorkflowRun` on that class".

Also, this is consistent with Python and Java and PHP, the other SDKs that use language metadata to mark workflow entry points.